pep-3157.txt

```diff
@@ -0,0 +1,247 @@
+PEP: 3157
+Title: Syntactic formalization of f-strings
+Version: $Revision$
+Last-Modified: $Date$
+Author: Pablo Galindo <pablogsal@python.org>,
+        Batuhan Taskaya <batuhan@python.org>,
+        Lysandros Nikolaou <lisandrosnik@gmail.com>
+Sponsor: Pablo Galindo <pablogsal@python.org>
+Discussions-To: python-dev@python.org
+Type: Standards Track
+Status: Draft
+Content-Type: text/x-rst
+Python-Version: 3.9
+Created: 15-Nov-2021
+
+
+Abstract
+========
+
+This document proposes to lift some of the restrictions originally formulated in
+:pep:`498` and to provide a formalized grammar for f-strings that can be
+integrated into parser directly. The proposed syntactic formalization of
+f-strings will have some small side-effects over how f-strings are parsed and
+interpreted, allowing for a considerable number of advantages for end users and
+library developers, while also improving dramatically the maintainance cost of
+the code dedicated to parse f-strings.
+
+
+Motivation
+==========
+
+When f-strings were originally introduced in :pep:`398` the specification was
+provided without providing a formal Grammar for f-strings. Additionally, theGG
+specification contains several restrictions that are imposed so the parsing of
+f-strings could be implemented into CPython without modifying the existing
+lexer. These limitations have been recognized previously and previous attempt
+have been made to lift them in :pep:`536` but non of this work was ever implemented 
+(see [2]_). Some of these limitations (collected originally by :pep:`536`) are:
+
+#. It is impossible to use the quote character delimiting the f-string
+   within the expression portion::
+
+    >>> f'Magic wand: { bag['wand'] }'
+                                 ^
+    SyntaxError: invalid syntax
+
+#. A previously considered way around it would lead to escape sequences
+   in executed code and is prohibited in f-strings::
+
+    >>> f'Magic wand { bag[\'wand\'] } string'
+    SyntaxError: f-string expression portion cannot include a backslash
+
+#. (TODO: DO WE WANT THIS?) Comments are forbidden even in multi-line f-strings::
+
+    >>> f'''A complex trick: {
+    ... bag['bag']  # recursive bags!
+    ... }'''
+    SyntaxError: f-string expression part cannot include '#'
+
+#. (TODO: DO WE WANT THIS?) Expression portions need to wrap ``':'`` and ``'!'`` in braces::
+
+    >>> f'Useless use of lambdas: { lambda x: x*2 }'
+    SyntaxError: unexpected EOF while parsing
```

Comment:
I'm somewhat torn on this, I feel that if we present the lambda as the example, support will be for not allowing it, since defining (and calling) a lambda in an f-string does not present a use case that makes sense. Are there any other examples we could use maybe?

Reply:
I added it here to discuss it together but I think this is cursed. One of the reasons is that the : defined the format specifier and we need to parse that as a string but if we allow that chunk to be tokenised normally it will never match sone of the allowed format specifiers because those are invalid combinations of alphanumeric characters.

This means that the tokenizer will always emit a FSTRING_MIDDLE and the lambda will never parse.

Reply:
@isidentical @lysnikolaou thoughts?

Reply:
> One of the reasons is that the : defined the format specifier and we need to parse that as a string but if we allow that chunk to be tokenised normally it will never match sone of the allowed format specifiers because those are invalid combinations of alphanumeric characters.

I guess that's true. I was thinking about how we can circumvent it, but without a direct feedback (or at least coupling with) from the parser (which I know we don't want to do) I don't see a way around it. I also agree with @lysnikolaou that this would not be a particularly good example, so I'd be inclined to change this example (or drop it).

Reply:
I guess we could do something like what we do with the f-string debug expressions and grab the source directly even if is a lot of random tokens and then build a string with that. We would need a special parser rule than can eat any number of tokens until a given one, which is possible but I don’t know if is worth it

Reply:
> I guess we could do something like what we do with the f-string debug expressions and grab the source directly even if is a lot of random tokens and then build a string with that.

That's true too! I didn't think about it purely because I was thinking about unlike debug expressions the format spec is not a formal python expression.

> We would need a special parser rule than can eat any number of tokens until a given one, which is possible but I don’t know if is worth it

In the event that we also formalize the ['format specification mini-language'](https://docs.python.org/3/library/string.html#formatspec) right in the parser for f-strings, it might be worth it (so we wouldn't do weird any-token eater rules) but I'd recommend it as a subsequent (and minor) improvement which bases itself on this. 

My vote is for deferring this case (particularly where a top-level colon is used, [assuming this doesn't affect colon-equal?]) for this PEP to the future with a recommendation/mention that it is possible to do but might require either the any-token approach you mentioned or a formalization of the format spec language in the grammar which would be a separate feature that'd need this PEP.

Reply:
> We would need a special parser rule than can eat any number of tokens until a given one, which is possible but I don’t know if is worth it

I'm probably also for "it's not worth it", so dropping this altogether (at least for now) makes sense to me.

> In the event that we also formalize the ['format specification mini-language'](https://docs.python.org/3/library/string.html#formatspec) right in the parser for f-strings, it might be worth it (so we wouldn't do weird any-token eater rules) but I'd recommend it as a subsequent (and minor) improvement which bases itself on this.

I would be +1 on this exactly as presented. We can start off with what we have now and improve over it later.

Reply:
Ok, let's move it then to "rejected ideas" or something similar. Are you ok with that?

Reply:
Agreed.

Reply:
Yup, let's move to rejected ideas.

Reply:
Done